### PR TITLE
Fix/387

### PR DESCRIPTION
**Pull request by evilaliv3**
This commit fixes issues and doubts as by ticket https://github.com/openpgpjs/openpgpjs/issues/387

The pull request deprecate https://github.com/openpgpjs/openpgpjs/pull/388 of which cherry-picks few commits.

@tanx: seems it is your turn my friend!

### CODE REVIEW

**Comment by evilaliv3** [comment]
@tanx but why you say it does not fix the issue?

doesnt this work for you?
```bash
git clone git@github.com:openpgpjs/openpgpjs.git
cd openpgpjs
git checkeout fix/387
npm install -d
grunt
grunt test
```

the same the tests on travis are working: https://travis-ci.org/openpgpjs/openpgpjs/builds/104202739

do not mind to the browsertesting that has the tunnels not working due of too much test runned, but in few minuts i will restart them.

**Comment by tanx** [comment]
See my explanation [here](https://github.com/openpgpjs/openpgpjs/issues/387#issuecomment-174082166). I've been using this workaround with pointing directly to `../../dist/openpgp/` as well because I have not been able to identify why the `'openpgp'` module cannot be resolved locally anymore. Like I said in the linked explanation. I'm pretty sure the answer lies in the difference between the cached node_modules folder in travis and our local installatiom after a clean `npm install`.

Do you if there is anyway to download the contents of a travis cache?

**Comment by evilaliv3** [comment]
not an official way, but eventually this one: https://github.com/evilaliv3/travis-ci-reverse-shell :)

File: Gruntfile.js [hunk 1]
```diff
@@ -228,32 +226,13 @@ module.exports = function(grunt) {
...
 
   grunt.registerTask('documentation', ['jsdoc']);
 
-  // Alias the `npm_pack` task to run `npm pack`
-  grunt.registerTask('npm_pack', 'npm pack', function () {
-    var done = this.async();
-    var npm = require('child_process').exec('npm pack ../', { cwd: 'dist'}, function (err, stdout) {
-      var package = stdout;
-      if (err === null) {
-        var install = require('child_process').exec('npm install dist/' + package, function (err) {
```
Comment (tanx) [inline Gruntfile.js]: Aha! It seems the travis cache just had an installation of openpgp from the previous build 😄

We probably had this in our local node_modules folder as well after the first successful grunt buil 😉

**Comment by evilaliv3** [comment]
anyway i got all your points, and i'm sure you are right but it's also true that the npm_pack is hiding this possible bug of packaging since a lot.

to find the breaking point for that we shall do a bisect over the commits using this test case:
```
npm install -d
grunt
rm node_modules/openpgp
grunt test
```

an we shall find the in which we do not get an error

File: Gruntfile.js [hunk 2]
```diff
@@ -48,9 +48,7 @@ module.exports = function(grunt) {
           'test/lib/unittests-bundle.js': [ './test/unittests.js' ]
         },
         options: {
-          browserifyOptions: {
-            external: [ 'openpgp', 'crypto', 'node-localstorage']
-          }
+          external: [ 'openpgp']
```
Comment (tanx) [inline Gruntfile.js]: We should keep 'crypto' 'node-localstorage' external for the browser unit tests. Any reason you added them here?

**Comment by evilaliv3** [comment]
@tanx: for a matter of curiosity i retested on a clean v1.2.0 removing the nodemodules/openpgp and i get the same issue there.

File: Gruntfile.js (hunk 2 above)
Comment (evilaliv3) [inline Gruntfile.js]: do you mean why i removed them here (you said added)?

i was about to tell you, i removed them as they are included as external by 'openpgp'; am i wrong?

File: Gruntfile.js (hunk 1 above)
Comment (evilaliv3) [inline Gruntfile.js]: exactly :P this is what i'm trying to repeat to you in all the varius pull :)

sorry for the confusion of today; i'm also tired but now i think you can understand what i was saying here and there

File: Gruntfile.js (hunk 2 above)
Comment (evilaliv3) [inline Gruntfile.js]: ah! now i understand why you are talking in reversed;
by the way the openpgpjs module automatically; probably yes, this line should be  "external: [ 'openpgp', 'crypto', 'node-localstorage']" but i let this to you as you have more experience in this.

**Comment by tanx** [comment]
> @tanx: for a matter of curiosity i retested on a clean v1.2.0 removing the nodemodules/openpgp and i get the same issue there.

Yeah I think you're right. I guess it is not normal behavior for node to resolve 'openpgp' using the project's own package.json. I was only assuming this since there was a local installation hiding under node_modules/openpgp which resulted from the npm_pack grunt job.

Anyway it seems your fix seems to work. BUT we need to keep in mind that we need to test dist/openpgp.js instead of src/index.js because of the babelify build. So let me try to use that locally instead get back to you with a suggestion after testing that locally.

**Comment by evilaliv3** [comment]
dist/openpgp.js is tested in all browsers tests: https://github.com/openpgpjs/openpgpjs/blob/master/test/unittests.html

instread for tracking code coverage we run tests on sources

this mixed approach shoould work i think

**Comment by tanx** [comment]
> this mixed approach shoould work i think

Yes it will right now. But as soon as we have ES6 syntax in src/ it will break e.g. for node v0.12. See my PR here: https://github.com/openpgpjs/openpgpjs/pull/390

**Comment by evilaliv3** [comment]
perfect! :) so looking forward for this mix:

ES6 + ArrayBuffers + streaming patch for encryption + updated selection of crypto libraries

i'm dreaming of this but in the short term i would need from you simple the review of streaming patch that is really selfcontained and unittested; aside from strong passion this is why i'm trying to put all my effort in support you when a bug pops out. :)

**Comment by tanx** [comment]
> ES6 + ArrayBuffers + streaming patch for encryption + updated selection of crypto libraries
> 
> i'm dreaming of this but in the short term i would need from you simple the review of streaming patch that is really selfcontained and unittested; aside from strong passion this is why i'm trying to put all my effort in support you when a bug pops out. :)

I hear you! That's why I'm keeping the diff with ES6 on it's own branch for now. The use of `dist/openpgp` is just to stay future proof. Nothing more nothing less. Having said that, I will probably get back to the ES6 branch after reviewing/merging the ArrayBuffers and streaming patch for encryption PRs. I will probably look at ArrayBuffers first though.

**Comment by evilaliv3** [comment]
perfect for me to proceed on the array buffer this way i wil support you on the retesting cause i'm also interested in the possible performance/memory improvement that may follow.

in the last days i did a little of review of it but aside from requiring to be cleaned form minor mistakes and rebased is a simple mapping one-to-one.

**Pull request closed without merging**